diff --git a/lib/aws-sdk-core/plugins/logging.rb b/lib/aws-sdk-core/plugins/logging.rb
index 1680232a9..c91a3a1fb 100644
--- a/lib/aws-sdk-core/plugins/logging.rb
+++ b/lib/aws-sdk-core/plugins/logging.rb
@@ -50,7 +50,7 @@ is not set, logging will be disabled.
         # @param [Response] response
         # @return [void]
         def log(config, response)
-          config.logger.send(config.log_level, format(config, response))
+          config.logger.send(:debug, format(config, response))
         end
 
         # @param [Configuration] config
