--- systemd-journal/systemd-journal.gemspec.orig
+++ systemd-journal/systemd-journal.gemspec
@@ -15,11 +15,6 @@ Gem::Specification.new do |gem|
 
   gem.cert_chain    = Dir['certs/john@throttle.io.pem']
 
-  if $PROGRAM_NAME.end_with?('gem')
-    gem.signing_key = ENV['GEM_SIGNING_KEY']
-    raise ArgumentError, 'Please set GEM_SIGNING_KEY' if gem.signing_key.nil?
-  end
-
   gem.files         = `git ls-files`.split($/)
   gem.executables   = gem.files.grep(%r{^bin/}).map { |f| File.basename(f) }
   gem.test_files    = gem.files.grep(%r{^(test|spec|features)/})
