--- systemd-journal/systemd-journal.gemspec.orig	2019-02-21 16:22:08.609602292 -0700
+++ systemd-journal/systemd-journal.gemspec	2019-02-21 16:22:15.576604622 -0700
@@ -14,13 +14,8 @@
   gem.homepage      = 'https://github.com/ledbettj/systemd-journal'
 
   gem.cert_chain    = Dir['certs/*']
 
-  if $PROGRAM_NAME.end_with?('gem')
-    gem.signing_key = ENV['GEM_SIGNING_KEY']
-    raise ArgumentError, 'Please set GEM_SIGNING_KEY' if gem.signing_key.nil?
-  end
-
   gem.files         = `git ls-files`.split($/)
   gem.executables   = gem.files.grep(%r{^bin/}).map { |f| File.basename(f) }
   gem.test_files    = gem.files.grep(%r{^(test|spec|features)/})
   gem.require_paths = ['lib']
