diff -up ./rsyslog.service.in.service ./rsyslog.service.in
--- ./rsyslog.service.in.service	2018-03-01 13:58:11.480598935 +0100
+++ ./rsyslog.service.in	2018-03-01 13:58:25.433518607 +0100
@@ -1,12 +1,16 @@
 [Unit]
 Description=System Logging Service
 Requires=syslog.socket
+Wants=network.target network-online.target
+After=network.target network-online.target
 Documentation=man:rsyslogd(8)
 Documentation=http://www.rsyslog.com/doc/
 
 [Service]
 Type=notify
-ExecStart=@sbindir@/rsyslogd -n -iNONE
+EnvironmentFile=-/etc/sysconfig/rsyslog
+ExecStart=@sbindir@/rsyslogd -n $SYSLOGD_OPTIONS
+UMask=0066
 StandardOutput=null
 Restart=on-failure
 
