// infra/server/lib/adapters/metrics/adapter_types.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var InfraMetricModelMetricType;
(function (InfraMetricModelMetricType) {
    InfraMetricModelMetricType["avg"] = "avg";
    InfraMetricModelMetricType["max"] = "max";
    InfraMetricModelMetricType["min"] = "min";
    InfraMetricModelMetricType["calculation"] = "calculation";
    InfraMetricModelMetricType["cardinality"] = "cardinality";
    InfraMetricModelMetricType["series_agg"] = "series_agg";
    InfraMetricModelMetricType["positive_only"] = "positive_only";
    InfraMetricModelMetricType["derivative"] = "derivative";
    InfraMetricModelMetricType["count"] = "count";
})(InfraMetricModelMetricType = exports.InfraMetricModelMetricType || (exports.InfraMetricModelMetricType = {}));
