// infra/public/pages/metrics/layouts/types.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var InfraMetricLayoutVisualizationType;
(function (InfraMetricLayoutVisualizationType) {
    InfraMetricLayoutVisualizationType["line"] = "line";
    InfraMetricLayoutVisualizationType["area"] = "area";
    InfraMetricLayoutVisualizationType["bar"] = "bar";
})(InfraMetricLayoutVisualizationType = exports.InfraMetricLayoutVisualizationType || (exports.InfraMetricLayoutVisualizationType = {}));
var InfraMetricLayoutSectionType;
(function (InfraMetricLayoutSectionType) {
    InfraMetricLayoutSectionType["chart"] = "chart";
    InfraMetricLayoutSectionType["gauges"] = "gauges";
})(InfraMetricLayoutSectionType = exports.InfraMetricLayoutSectionType || (exports.InfraMetricLayoutSectionType = {}));
