// infra/public/pages/logs/logs.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
var _a;
"use strict";
const eui_1 = require("@elastic/eui");
const react_1 = require("@kbn/i18n/react");
const react_2 = tslib_1.__importDefault(require("react"));
const page_content_1 = require("./page_content");
const toolbar_1 = require("./toolbar");
const no_indices_1 = require("../../components/empty_states/no_indices");
const header_1 = require("../../components/header");
const log_flyout_1 = require("../../components/logging/log_flyout");
const page_1 = require("../../components/page");
const header_feedback_link_1 = require("../../components/header_feedback_link");
const source_configuration_1 = require("../../components/source_configuration");
const source_configuration_flyout_state_1 = require("../../components/source_configuration/source_configuration_flyout_state");
const with_log_filter_1 = require("../../containers/logs/with_log_filter");
const with_log_flyout_1 = require("../../containers/logs/with_log_flyout");
const with_log_flyout_options_1 = require("../../containers/logs/with_log_flyout_options");
const with_log_flyout_options_2 = require("../../containers/logs/with_log_flyout_options");
const with_log_minimap_1 = require("../../containers/logs/with_log_minimap");
const with_log_position_1 = require("../../containers/logs/with_log_position");
const with_log_textview_1 = require("../../containers/logs/with_log_textview");
const with_kibana_chrome_1 = require("../../containers/with_kibana_chrome");
const with_source_1 = require("../../containers/with_source");
exports.LogsPage = react_1.injectI18n((_a = class extends react_2.default.Component {
        render() {
            const { intl } = this.props;
            return (react_2.default.createElement(page_1.ColumnarPage, null,
                react_2.default.createElement(header_1.Header, { appendSections: react_2.default.createElement(header_feedback_link_1.InfraHeaderFeedbackLink, { url: "https://discuss.elastic.co/c/logs" }), breadcrumbs: [
                        {
                            text: intl.formatMessage({
                                id: 'xpack.infra.logsPage.logsBreadcrumbsText',
                                defaultMessage: 'Logs',
                            }),
                        },
                    ] }),
                react_2.default.createElement(with_source_1.WithSource, null, ({ derivedIndexPattern, hasFailed, isLoading, lastFailureMessage, load, logIndicesExist, sourceId, }) => (react_2.default.createElement(react_2.default.Fragment, null,
                    react_2.default.createElement(source_configuration_1.SourceConfigurationFlyout, null),
                    isLoading ? (react_2.default.createElement(with_source_1.SourceLoadingPage, null)) : logIndicesExist ? (react_2.default.createElement(react_2.default.Fragment, null,
                        react_2.default.createElement(with_log_filter_1.WithLogFilterUrlState, { indexPattern: derivedIndexPattern }),
                        react_2.default.createElement(with_log_position_1.WithLogPositionUrlState, null),
                        react_2.default.createElement(with_log_minimap_1.WithLogMinimapUrlState, null),
                        react_2.default.createElement(with_log_textview_1.WithLogTextviewUrlState, null),
                        react_2.default.createElement(with_log_flyout_options_2.WithFlyoutOptionsUrlState, null),
                        react_2.default.createElement(toolbar_1.LogsToolbar, null),
                        react_2.default.createElement(with_log_filter_1.WithLogFilter, { indexPattern: derivedIndexPattern }, ({ applyFilterQueryFromKueryExpression }) => (react_2.default.createElement(react_2.default.Fragment, null,
                            react_2.default.createElement(with_log_flyout_options_1.WithFlyoutOptions, null, ({ showFlyout, setFlyoutItem }) => (react_2.default.createElement(page_content_1.LogsPageContent, { showFlyout: showFlyout, setFlyoutItem: setFlyoutItem }))),
                            react_2.default.createElement(with_log_flyout_1.WithLogFlyout, { sourceId: sourceId }, ({ flyoutItem, hideFlyout, loading }) => (react_2.default.createElement(log_flyout_1.LogFlyout, { setFilter: applyFilterQueryFromKueryExpression, flyoutItem: flyoutItem, hideFlyout: hideFlyout, loading: loading })))))))) : hasFailed ? (react_2.default.createElement(with_source_1.SourceErrorPage, { errorMessage: lastFailureMessage || '', retry: load })) : (react_2.default.createElement(with_kibana_chrome_1.WithKibanaChrome, null, ({ basePath }) => (react_2.default.createElement(no_indices_1.NoIndices, { title: intl.formatMessage({
                            id: 'xpack.infra.logsPage.noLoggingIndicesTitle',
                            defaultMessage: "Looks like you don't have any logging indices.",
                        }), message: intl.formatMessage({
                            id: 'xpack.infra.logsPage.noLoggingIndicesDescription',
                            defaultMessage: "Let's add some!",
                        }), actions: react_2.default.createElement(eui_1.EuiFlexGroup, null,
                            react_2.default.createElement(eui_1.EuiFlexItem, null,
                                react_2.default.createElement(eui_1.EuiButton, { href: `${basePath}/app/kibana#/home/tutorial_directory/logging`, color: "primary", fill: true }, intl.formatMessage({
                                    id: 'xpack.infra.logsPage.noLoggingIndicesInstructionsActionLabel',
                                    defaultMessage: 'View setup instructions',
                                }))),
                            react_2.default.createElement(eui_1.EuiFlexItem, null,
                                react_2.default.createElement(source_configuration_flyout_state_1.WithSourceConfigurationFlyoutState, null, ({ enable }) => (react_2.default.createElement(eui_1.EuiButton, { color: "primary", onClick: enable }, intl.formatMessage({
                                    id: 'xpack.infra.configureSourceActionLabel',
                                    defaultMessage: 'Change source configuration',
                                })))))) })))))))));
        }
    },
    _a.displayName = 'LogsPage',
    _a));
