// x-pack/plugins/uptime/server/lib/adapters/monitors/memory_pings_adapter.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
class UMMemoryMonitorsAdapter {
    constructor(monitorsDB) {
        this.monitorsDB = monitorsDB;
    }
    async getLatestMonitors(request, dateRangeStart, dateRangeEnd, filters) {
        return this.monitorsDB.filter(ping => {
            // const timestamp = moment(ping.timestamp).valueOf();
            throw new Error('Method not implemented.');
            // return dateRangeStart <= timestamp && timestamp <= dateRangeEnd;
        });
    }
    async getMonitorChartsData(req, monitorId, dateRangeStart, dateRangeEnd) {
        throw new Error('Method not implemented.');
    }
    async getSnapshotCount(request, dateRangeStart, dateRangeEnd, filters) {
        throw new Error('Method not implemented.');
    }
    async getFilterBar(request) {
        throw new Error('Method not implemented.');
    }
    async getErrorsList(request, dateRangeStart, dateRangeEnd, filters) {
        throw new Error('Method not implemented.');
    }
}
exports.UMMemoryMonitorsAdapter = UMMemoryMonitorsAdapter;
