// infra/public/store/actions.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var local_1 = require("./local");
exports.logFilterActions = local_1.logFilterActions;
exports.logMinimapActions = local_1.logMinimapActions;
exports.logPositionActions = local_1.logPositionActions;
exports.logTextviewActions = local_1.logTextviewActions;
exports.metricTimeActions = local_1.metricTimeActions;
exports.waffleFilterActions = local_1.waffleFilterActions;
exports.waffleTimeActions = local_1.waffleTimeActions;
exports.waffleOptionsActions = local_1.waffleOptionsActions;
exports.flyoutOptionsActions = local_1.flyoutOptionsActions;
var remote_1 = require("./remote");
exports.logEntriesActions = remote_1.logEntriesActions;
exports.logSummaryActions = remote_1.logSummaryActions;
